### PR TITLE
[#166] Integrate Bootable JAR application images

### LINKED ISSUE #166: As an administrator, I can deploy an application image using the bootable jar with the WildFly Operator
## Overview
The EAP Operator input is an application image. Until now, the requirement was that this application image was built using EAP S2I builder and runtime images. Since WildFly 21, the Bootable JAR (also known as “fat JAR”) is a new way to package a WildFly server and an application both in an executable JAR. 

The goal of this RFE is to add the ability to deploy a Bootable JAR application image by using the WildFly Operator.

## Dev Contacts
Yeray Borges (yborgess@redhat.com)
## QE Contacts
Jan Blizňák (jbliznak@redhat.com)
## Testing By
QE
## Affected Projects or Components
[WildFly Operator](https://github.com/wildfly/wildfly-operator)
## Other Interested Projects
N/A
### Hard Requirements

- Bootable JAR application image requirements:
   - The base image for the Bootable JAR container must be registry.access.redhat.com/ubi8/openjdk-11 or any other Red Hat ubi8 supplying a higher JDK version.
   - The Bootable JAR must be built for a server configured for the cloud environment. This is done by enabling the <cloud> configuration item on the wildfly-jar-maven-plugin configuration. See https://docs.wildfly.org/bootablejar/#wildfly_jar_configuring_cloud
   - For example, one way to generate a Bootable JAR application image from an existing Bootable JAR project is by using the [JKube Kubernetes maven plugin ](https://www.eclipse.org/jkube/docs/kubernetes-maven-plugin) . The following is an example of its uses in a maven project:
 ```
<project xmlns="http://maven.apache.org/POM/4.0.0" ...>
   ...
    <properties>
           ...
        <jkube.generator.from>registry.redhat.io/ubi8/openjdk-11:latest</jkube.generator.from>
    </properties>

    <dependencies>
        ...
    </dependencies>

    <build>
        <finalName>${project.artifactId}</finalName>
        <plugins>
            ...
            <plugin>
                <groupId>org.wildfly.plugins</groupId>
                <artifactId>wildfly-jar-maven-plugin</artifactId>
                <version>${version.bootable.jar}</version>
                <configuration>
                    <feature-pack-location>wildfly@maven(org.jboss.universe:community-universe)#${version.wildfly}</feature-pack-location>
                    <layers>
                        <layer>...</layer>
                    </layers>
                    <cloud/>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>package</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.eclipse.jkube</groupId>
                <artifactId>kubernetes-maven-plugin</artifactId>
                <version>${version.jkube}</version>
            </plugin>
        </plugins>
    </build>
</project>
```
Upon the above pom.xml example, the Application Bootable JAR image can be generated by issuing the following maven goals:
`mvn package k8s:build`
The `kubernetes-maven-plugin` will create a Dockerfile available at `${project.build.directory}/target/docker`. This image is also tagged and made available on your local docker registry.
- Custom Resource configuration requirements
   - The “applicationImage” attribute will be used to identify the Bootable JAR application image. It will accept the same formats already available for S2I application images:
     - An image name: quay.io/bootable-jar/wildfly-demo
     - An image tag: quay.io/bootable-jar/wildfly-demo:1.0
     - An image hash: quay.io/bootable-jar/wildfly-demo@sha256:47c06c96e80d0defb777686cdb468c636d9b3b7081a35b784330a050a403e15b
     - An OpenShift image stream: wildfly-demo:latest
   - The WildFly operator CRD will supply a new boolean configuration attribute named “bootable-jar”, defaulted to false, to identify whether the application image is a Bootable JAR or if it is a S2I application image.
      - This value must be ‘true’ to enable the Bootable JAR images.
- Transaction Recovery facility is not supported on the Bootable JAR. 
    - The WildFly Operator will not run the Transaction Recovery facility when a pod running a Bootable JAR application image is scaled down.
    - The WildFly Operator will log a trace describing this situation when a pod is scaled down.
- Supply custom server configuration file via config map is not supported a Bootable JAR application image.
    - The natural way to proceed with a Bootable JAR is to configure the server via CLI script when the Bootable JAR is created.
- The WildFly Operator will inject the liveness and readiness probes. The injected probes will be:
   - liveness: An HTTPGet probe that checks the /health/live resource
   - readiness: An HTTPGet probe that checks /health/ready resource

### Nice-to-Have Requirements
N/A

### Non-Requirements
N/A

## Test Plan
- An example Custom Resource using a Bootable JAR image will be supplied under the deploy/crds directory. This example CR can be used to run a manual test and verify the Operator deploys the image in an OpenShift cluster.
- The Operator test suite will include a test case which will deploy a Bootable JAR application image to verify the required replicas are created without errors.

## Community Documentation
The user guide documentation and the WildFlyServer CRD documentation will involve to reflect the changes introduced by this RFE

## Release Note Content
Added the ability to deploy a Bootable JAR image.

### PR DESCRIPTION
**Pull request by yersan**
Fixes #166

### CODE REVIEW

File: deploy/crds/quickstart-bootable-jar-cr.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+apiVersion: wildfly.org/v1alpha1
+kind: WildFlyServer
+metadata:
+  name: quickstart-bootable-jar
+spec:
+  applicationImage: "quay.io/wildfly-quickstarts/wildfly-operator-quickstart:bootable-21.0"
```
Comment (yersan) [inline deploy/crds/quickstart-bootable-jar-cr.yaml:6]: @jmesnil This image does not exist yet, you could tag and push to wildfly-quickstarts the onw I've been using quay.io/yborgess/jaxrs-demo:1.0. 

The quay.io/wildfly-quickstarts/wildfly-operator-quickstart:bootable-21.0 is also needed for the test.

Comment (jmesnil) [reply deploy/crds/quickstart-bootable-jar-cr.yaml:6]: done, pushed to quay.io/wildfly-quickstarts/wildfly-operator-quickstart:bootable-21.0

**Review by jmesnil — APPROVED** [review]

**Comment by jmesnil** [comment]
@yersan thanks!

**Pull request merged**